libbeat/processors/cache/file_store.go

```diff
@@ -0,0 +1,262 @@
+	}
+	s.readState()
+	return &s
+}
+
+func (c *fileStore) String() string { return "file:" + c.id }
+
+// dropFrom decreases the reference count for the fileStore and removes it from
+// the stores map if the count is zero. dropFrom is safe for concurrent use.
+func (c *fileStore) dropFrom(stores *fileStoreSet) {
+	c.mu.Lock()
+	c.refs--
+	if c.refs < 0 {
+		panic("invalid reference count")
+	}
+	if c.refs == 0 {
+		// Stop periodic writes
+		c.cancel()
+		// and do a final write out.
+		c.writeState(true)
+
+		stores.free(c.id)
+		// GC assists.
+		c.cache = nil
+		c.expiries = nil
+	}
+	c.mu.Unlock()
+}
+
+func (c *fileStore) readState() {
+	f, err := os.Open(c.path)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			c.log.Debugw("no state on file system", "error", err)
+		} else {
+			c.log.Errorw("failed to open file to read state", "error", err)
+		}
+		return
+	}
+	defer f.Close()
+
+	// It would be nice to be able at this stage to determine
+	// whether the file is stale past the TTL of the cache, but
+	// we do not have this information yet. So we must read
+	// through all the elements. If any survive the filter, we
+	// were alive, otherwise delete the file.
+
+	dec := json.NewDecoder(f)
+	for {
+		var e CacheEntry
+		err = dec.Decode(&e)
+		if err != nil {
+			if err != io.EOF {
+				c.log.Errorw("failed to read state element", "error", err)
+			}
+			break
+		}
+		if e.Expires.Before(time.Now()) {
+			// Don't retain expired elements.
+			continue
+		}
+		c.cache[e.Key] = &e
+		heap.Push(&c.expiries, &e)
+	}
+
+	if len(c.cache) != 0 {
+		return
+	}
+	// We had no live entries, so delete the file.
+	err = os.Remove(c.path)
+	if err != nil {
+		c.log.Errorw("failed to delete stale cache file", "error", err)
+	}
+}
+
+// periodicWriteOut writes the cache contents to the backing file at the
+// specified interval until the context is cancelled. periodicWriteOut is
+// safe for concurrent use.
+func (c *fileStore) periodicWriteOut(ctx context.Context, every time.Duration) {
+	tick := time.NewTicker(every)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			c.mu.Lock()
+			c.writeState(false)
+			c.mu.Unlock()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// writeState writes the current cache state to the backing file.
+// If final is true and the cache is empty, the file will be deleted.
+func (c *fileStore) writeState(final bool) {
+	if len(c.cache) == 0 && final {
+		err := os.Remove(c.path)
+		if err != nil {
+			c.log.Errorw("failed to delete write state when empty", "error", err)
+		}
+		return
+	}
+	f, err := os.CreateTemp("", "")
+	if err != nil {
+		c.log.Errorw("failed to open file to write state", "error", err)
+		return
+	}
+	// We are writing into tmp, so make sure we are private.
+	err = os.Chmod(f.Name(), 0o600)
+	if err != nil {
+		c.log.Errorw("failed to set state file mode", "error", err)
+		return
+	}
+	tmp := f.Name()
+	defer func() {
+		err = f.Close()
```

Comment:
One thing to be aware of is that write operations are cached on Windows. So even though this may have closed the file, it's not guaranteed to be synced to disk. You could incorporate an `f.Sync()` call to ensure this is flush from the write cache to disk (it does a https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-flushfilebuffers).

Reply:
This is theoretically true on posix as well (I was considering this, but didn't do it).